internal/enginenetx/DESIGN.md

```diff
@@ -0,0 +1,815 @@
+      }
+    }
+  }
+  "Version": 5
+}
+```
+
+**Listing 7.** Content of the stats state as cached on disk.
+
+That is, the `DomainEndpoints` map contains contains an entry for each
+TLS endpoint and, in turn, such an entry contains tactics indexed by
+a summary string to speed up looking them up.
+
+For each tactic, we keep counters and histograms, the time when the
+entry had been updated last, and the tactic itself.
+
+The `*statsManager` implements `httpsDialerEventsHandler`, which means
+that it has callbacks invoked by the `*httpsDialer` for interesting
+events regarding dialing (e.g., whether TCP connect failed).
+
+These callbacks basically create or update stats by locking a mutex
+and updating the relevant counters and histograms.
+
+## Real-World Scenarios
+
+This section illustrates the behavior of this package under specific
+network failure conditions, with specific emphasis on what happens if
+the bridge IP address becomes, for any reason, unavailable. (We are
+doing this because all this work was prompeted by addressing the
+[ooni/probe#2704](https://github.com/ooni/probe/issues/2704) issue.)
+
+### Invalid bridge without cached data
+
+In this first scenario, we're showing what happens if the bridge IP address
+becomes unavailable without any previous state saved on disk. (To emulate
+this scenario, change the bridge IP address in [bridgespolicy.go](bridgespolicy.go)
+to become `10.0.0.1`, recompile, and wipe `httpsdialerstats.state`).
+
+Here's an excerpt from the logs:
+
+```
+[      0.001346] <info> httpsDialer: [#1] TCPConnect 10.0.0.1:443... started
+[      0.002101] <info> sessionresolver: lookup api.ooni.io using https://wikimedia-dns.org/dns-query... started
+[      0.264132] <info> sessionresolver: lookup api.ooni.io using https://wikimedia-dns.org/dns-query... ok
+[      0.501774] <info> httpsDialer: [#1] TCPConnect 10.0.0.1:443... in progress
+[      1.002330] <info> httpsDialer: [#2] TCPConnect 10.0.0.1:443... started
+[      1.503687] <info> httpsDialer: [#2] TCPConnect 10.0.0.1:443... in progress
+[      2.001488] <info> httpsDialer: [#4] TCPConnect 162.55.247.208:443... started
+[      2.046917] <info> httpsDialer: [#4] TCPConnect 162.55.247.208:443... ok
+[      2.047016] <info> httpsDialer: [#4] TLSHandshake with 162.55.247.208:443 SNI=api.ooni.io ALPN=[h2 http/1.1]... started
+[      2.093148] <info> httpsDialer: [#4] TLSHandshake with 162.55.247.208:443 SNI=api.ooni.io ALPN=[h2 http/1.1]... ok
+[      2.093181] <info> httpsDialer: [#4] TLSVerifyCertificateChain api.ooni.io... started
+[      2.095923] <info> httpsDialer: [#4] TLSVerifyCertificateChain api.ooni.io... ok
+[      2.096054] <info> httpsDialer: [#1] TCPConnect 10.0.0.1:443... interrupted
+[      2.096077] <info> httpsDialer: [#2] TCPConnect 10.0.0.1:443... interrupted
+```
+
+**Listing 8.** Run with no previous cached state and unreachable hardcoded bridge address.
+
+After 2s, we start dialing with the IP addresses obtained through the DNS.
```

Comment:
My understanding is that this is taking this amount of time, because we are trying the bridges first (see: https://github.com/ooni/probe-cli/pull/1552/files#r1576554009), can we invert the order so that we always try first via DNS in the case of an empty cache?

Reply:
Yes, it is possible to do that by changing the order with which we construct and compose policies.

I will modify the code to this effect.